File: python/src/falcon_core/math/domains/domain.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.domain import Domain as _CDomain

class Domain:
    """Python wrapper for Domain."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, min_val: Any, max_val: Any, lesser_bound_contained: Any, greater_bound_contained: Any) -> Domain:
        return cls(_CDomain.new(min_val, max_val, lesser_bound_contained, greater_bound_contained))

    @classmethod
    def from_json(cls, json: str) -> Domain:
        return cls(_CDomain.from_json(json))

    def lesser_bound(self, ) -> None:
        ret = self._c.lesser_bound()
        return ret

    def greater_bound(self, ) -> None:
        ret = self._c.greater_bound()
        return ret

    def lesser_bound_contained(self, ) -> None:
        ret = self._c.lesser_bound_contained()
        return ret

    def greater_bound_contained(self, ) -> None:
        ret = self._c.greater_bound_contained()
        return ret

    def contains(self, value: Any) -> None:
        ret = self._c.contains(value)
        return ret

    def get_range(self, ) -> None:
        ret = self._c.get_range()
        return ret

    def center(self, ) -> None:
        ret = self._c.center()
        return ret

    def intersection(self, other: Domain) -> Domain:
        ret = self._c.intersection(other._c if other is not None else None)
        return Domain._from_capi(ret)

    def union(self, other: Domain) -> Domain:
        ret = self._c.union(other._c if other is not None else None)
        return Domain._from_capi(ret)

    def is_empty(self, ) -> None:
        ret = self._c.is_empty()
        return ret

    def contains_domain(self, other: Domain) -> None:
        ret = self._c.contains_domain(other._c if other is not None else None)
        return ret

    def shift(self, offset: Any) -> Domain:
        ret = self._c.shift(offset)
        return Domain._from_capi(ret)

    def scale(self, scale: Any) -> Domain:
        ret = self._c.scale(scale)
        return Domain._from_capi(ret)

    def transform(self, other: Domain, value: Any) -> None:
        ret = self._c.transform(other._c if other is not None else None, value)
        return ret

    def equal(self, other: Domain) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: Domain) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Domain):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Domain):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/src/math/domains/domain.ml
open Ctypes
open Capi_bindings
open Error_handling

(* No opens needed - using qualified names *)

class type c_domain_t = object
  method raw : unit ptr
end
class c_domain (h : unit ptr) : c_domain_t = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.domain_destroy raw_val;
    Error_handling.raise_if_error ()
  ) self
end

module Domain = struct
  type t = c_domain

  let copy (handle : t) : t =
    Error_handling.read handle (fun () ->
      let ptr = Capi_bindings.domain_copy handle#raw in
      Error_handling.raise_if_error ();
      new c_domain ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.domain_from_json_string (Falcon_string.of_string json) in
    Error_handling.raise_if_error ();
    new c_domain ptr

  let make (min_val : float) (max_val : float) (lesser_bound_contained : bool) (greater_bound_contained : bool) : t =
    let ptr = Capi_bindings.domain_create min_val max_val lesser_bound_contained greater_bound_contained in
    Error_handling.raise_if_error ();
    new c_domain ptr

  let equal (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.domain_equal handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.domain_not_equal handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_to_json_string handle#raw in
      Error_handling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let lesserBound (handle : t) : float =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_lesser_bound handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let greaterBound (handle : t) : float =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_greater_bound handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let lesserBoundContained (handle : t) : bool =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_lesser_bound_contained handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let greaterBoundContained (handle : t) : bool =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_greater_bound_contained handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let in_ (handle : t) (value : float) : bool =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_in handle#raw value in
      Error_handling.raise_if_error ();
      result
    )

  let range (handle : t) : float =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_range handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let center (handle : t) : float =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_center handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let intersection (handle : t) (other : t) : t =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.domain_intersection handle#raw other#raw in
      Error_handling.raise_if_error ();
      new c_domain result
    )

  let union (handle : t) (other : t) : t =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.domain_union handle#raw other#raw in
      Error_handling.raise_if_error ();
      new c_domain result
    )

  let isEmpty (handle : t) : bool =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_is_empty handle#raw in
      Error_handling.raise_if_error ();
      result
    )

  let containsDomain (handle : t) (other : t) : bool =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.domain_contains_domain handle#raw other#raw in
      Error_handling.raise_if_error ();
      result
    )

  let shift (handle : t) (offset : float) : t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_shift handle#raw offset in
      Error_handling.raise_if_error ();
      new c_domain result
    )

  let scale (handle : t) (scale : float) : t =
    Error_handling.read handle (fun () ->
      let result = Capi_bindings.domain_scale handle#raw scale in
      Error_handling.raise_if_error ();
      new c_domain result
    )

  let transform (handle : t) (other : t) (value : float) : float =
    Error_handling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.domain_transform handle#raw other#raw value in
      Error_handling.raise_if_error ();
      result
    )

end